csrc/multidevice/device_mesh.cpp

```diff
@@ -56,4 +106,69 @@ int64_t DeviceMesh::size(const ParallelType parallel_type) const {
   return size();
 }
 
+std::vector<int64_t> DeviceMesh::getIndices(const DeviceIdxType device) const {
+  auto global_idx = idxOf(device);
+  if (global_idx == -1) {
+    return {};
+  }
+  std::vector<int64_t> indices(shape_.size());
+  int64_t accumulated_size = 1;
+  for (int64_t i = (int64_t)shape_.size() - 1; i >= 0; i--) {
+    indices[i] = (global_idx / accumulated_size) % shape_[i];
+    accumulated_size *= shape_[i];
+  }
+  return indices;
+}
+
+DeviceIdxType DeviceMesh::maxDeviceId() const {
+  return *std::max_element(vector_.begin(), vector_.end());
+}
+
+namespace {
+size_t ptypeToAxis(ParallelType ptype, size_t naxis) {
```

Comment:
```suggestion
int64_t ptypeToAxis(ParallelType ptype, size_t naxis) {
```
https://google.github.io/styleguide/cppguide.html#Integer_Types